Add secure IP propagation from teleport proxy to auth server via ALPN

AntonAM opened the pull request:
This PR adds a secure way to propagate information about client IP from proxy to the auth server through ALPN.
We're using PROXY protocol extensions called TLVs (Type-Length-Value) to send signed short lived JWT and proxy's certificate to the auth server. Auth validates JWT using provided signing certificate and host CA to make sure that IP information comes from our internal proxy. It also changes how we treat incoming PROXY headers - we still allow only one unsigned header, but we'll allow also one signed header coming from our proxies.
It will also be used as a base for adding IP propagation from proxy to the node server, so we can stop using unsafe `ProxyHelloSignature` way of doing it and for the PR to propagate IP to auth server when client connects not through ALPN but through SSH.

Fixes #15521

AntonAM committed:
Add secure IP propagation from proxy to auth server

We're using PROXY protocol extensions called TLVs to send
signed JWT and proxy's certificate to the auth server. Auth
validates JWT using provided signing certificate and host CA
to make sure that IP information comes from our internal proxy.

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

strideynet reviewed (commented):
The appropriate changes have been made for the ALPN Proxy - however - I believe you also need to add support for this into the SSH reverse tunnel proxy. See `lib/reversetunnel/transport.go` and follow the behaviour for when the address is the special constant `RemoteAuthServer` :)

strideynet commented on lib/jwt/jwt.go:
```diff
@@ -200,6 +200,29 @@ func (k *Key) SignSnowflake(p SignParams, issuer string) (string, error) {
+func (k *Key) SignPROXY(p PROXYSignParams) (string, error) {
+	claims := Claims{
+		Claims: jwt.Claims{
+			Subject:   fmt.Sprintf("%s/%s", p.SourceAddress, p.DestinationAddress),
```
Nit: I'd have a short function for producing the subject, just so we can prevent it drifting between the signing and verification side. It'll also provide us a place that'll be the source of truth.

AntonAM replied:
Added 👍

strideynet commented on lib/jwt/jwt.go:
```diff
@@ -200,6 +200,29 @@ func (k *Key) SignSnowflake(p SignParams, issuer string) (string, error) {
+func (k *Key) SignPROXY(p PROXYSignParams) (string, error) {
+	claims := Claims{
+		Claims: jwt.Claims{
+			Subject:   fmt.Sprintf("%s/%s", p.SourceAddress, p.DestinationAddress),
```
One wonders if it would make sense semantically for the subject to be `p.SourceAddress` and the audience to be `p.DestinationAddress`  ?

strideynet commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+		return *addr
+	}
+
+	parsedAddr := utils.FromAddr(a)
```
A comment here explaining in what situation we expect `a` to be something other than a `net.TCPAddr` would be helpful :)

AntonAM replied:
We don't really expect some specific situation that causes it, that's why I added comment about hot path. But technically it might be something else, that implements the interface 🤷

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+
+	foundRole := checkForSystemRole(identity, types.RoleProxy)
+	if !foundRole {
+		return false, nil
```
Maybe it would be helpful to emit a log message or an error here ! I imagine there's a chance a bad actor would try a certificate that does not belong to a proxy, and some visibility of this would be great for an operator/security team.

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+
+	identity, err := tlsca.FromSubject(signingCert.Subject, signingCert.NotAfter)
+	if err != nil {
+		return false, err
```
```suggestion
		return false, trace.Wrap(err)
```

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+	// Make sure that transmitted proxy cert is signed by one of the provided HostCAs
+	chains, err := signingCert.Verify(x509.VerifyOptions{Roots: roots})
+	if err != nil && !errors.As(err, &x509.UnknownAuthorityError{}) {
+		return false, err
```
```suggestion
		return false, trace.Wrap(err)
```

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+
+	signingCert, err := tlsca.ParseCertificatePEM(proxyCert)
+	if err != nil {
+		return false, err
```
```suggestion
		return false, trace.Wrap(err)
```

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+
+	token, proxyCert, err := p.getSignatureAndSigningCert()
+	if err != nil {
+		return false, err
```
```suggestion
		return false, trace.Wrap(err)
```

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+		RawToken:           token,
+	})
+	if err != nil {
+		return false, nil
```
Is this error meant to be propagated ?

strideynet commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
```
I wonder if we can get a context related to the connection here, rather than the one attached to the multiplexer ?

AntonAM replied:
At this moment it's basically just cancelation propagation from inner levels of multiplexer context, we don't really have any context related to the connection. Unless you have some idea?

strideynet replied:
This is probably fine tbh. It's a shame Go doesn't offer a context for connections out of the box.

strideynet commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
 			}
+
+			// If proxy line was signed but couldn't be verified it's ignored
+			if !verified && newProxyLine.isSigned() {
```
I think we might want to take a harsher stance here and reject the connection entirely - unless we imagine there's a normal operating scenario where this is likely to be signed but it could fail verification.

My fear being that if this fails and only logs, the user is likely to experience much more confusing errors (e.g if using IP pinning, they may end up having their connections rejected for coming from the wrong IP, they'd have to know this log message was related to their issue to work this out).

This harsher stance would also complicate things since we could make `VerifySignature` simply return an error rather than `(boolean, error)`

AntonAM replied:
Yes, I agree. I was on a fence about this, and probably taking a harsher stance makes sense 👍 

> This harsher stance would also **complicate** things 

Did you mean "simplify"? 😅

strideynet replied:
Urm, yes I did ! Might've reviewed this PR half asleep :')

AntonAM replied:
Changed it to the harsher stance logic.

strideynet commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+
+			// This is unsigned proxy line, return error if external PROXY protocol is not enabled
+			if !m.EnableExternalProxyProtocol {
+				return nil, trace.BadParameter("external proxy protocol support is disabled")
```
We should have consistency here - if we are willing to ignore an invalid signed PROXYv2, we should probably ignore and log rather than reject here. (But as explained in a previous comment, I think we should switch to rejecting an invalid signed PROXYv2)

AntonAM replied:
Yep, we're now failing unsuccessfully verified signed header as well.

strideynet commented on lib/multiplexer/proxyline.go:
```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+		if tlv.Type == PP2TypeTeleport {
+			teleportSubTLVs, err := UnmarshalTLVs(tlv.Value)
+			if err != nil {
+				return "", nil, err
```
```suggestion
		return "", nil, trace.Wrap(err)
```

AntonAM committed:
Use local cluster name instead of destination cluster

AntonAM committed:
Wrap returned error

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

AntonAM committed:
Wrap returned error

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

AntonAM committed:
Wrap returned error

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

AntonAM committed:
Wrap returned error

Co-authored-by: Noah Stride <noah.stride@goteleport.com>

AntonAM committed:
Wrap returned error

AntonAM commented:
> The appropriate changes have been made for the ALPN Proxy - however - I believe you also need to add support for this into the SSH reverse tunnel proxy. See `lib/reversetunnel/transport.go` and follow the behaviour for when the address is the special constant `RemoteAuthServer` :)

It's automatically supported already, since in the `func (s *AuthProxyDialerService) HandleConnection()` after calling `s.dialAuthServe()` we still just get `net.conn` which was created over the tunnel under the hood. And we write PROXY header to it in the same way.

Here's example of debug logs with two clusters (both on ip `192.168.2.23), and we're trying to list resources with `tsh ls` from a different IP (192.168.2.15), 

```
// Logs on the root cluster "tele2.port"
Dec  8 12:48:42.946884 [AUTH]      DEBU DBGG: Listing resources on Auth server, peer address: 192.168.2.15:57393 auth/auth_with_roles.go:1198
Dec  8 12:48:42.977950 [PROXY:SER] DEBU Requesting connection to @remote-auth-server [] in remote cluster. cluster:tele.port reversetunnel/remotesite.go:900

// Continuation of logs on the leaf cluster "tele.port"
Dec  8 12:48:42.982045 DEBU [PROXY:AGE] Successfully dialed to @remote-auth-server "", start proxying. cluster:tele2.port reversetunnel/transport.go:306
Dec  8 12:48:42.990815 DEBU [AUTH]      ClientCertPool -> cert(tele2.port issued by tele2.port:201604638341962741183117029050652578849) auth/middleware.go:676
Dec  8 12:48:42.990956 DEBU [AUTH]      ClientCertPool -> cert(tele2.port issued by tele2.port:209682280188038834517368231345398959698) auth/middleware.go:676
Dec  8 12:48:42.991061 DEBU [AUTH:1]    Server certificate cert(1b023d35-9484-48b3-b29d-788264c8de50.tele.port issued by tele.port:243357613536743763667664328504167968797). auth/middleware.go:312
Dec  8 12:48:43.019890 DEBU             Mapped roles [access editor] of remote user "teleport2-admin" to local roles [access] and traits map[aws_role_arns:[] db_names:[] db_users:[] kubernetes_groups:[] kubernetes_users:[] logins:[root ubuntu ec2-user -teleport-internal-join] windows_logins:[]]. services/access_checker.go:447
Dec  8 12:48:43.020701 DEBU             Skipping login -teleport-internal-join, not a valid Unix login. services/role.go:265
Dec  8 12:48:43.021186 DEBU [AUTH]      DBGG: Listing resources on Auth server, peer address: 192.168.2.15:57394 auth/auth_with_roles.go:1198
```

If I disable sending out PROXY header in `func (s *AuthProxyDialerService) HandleConnection()` we don't get correct client IP on the remote auth server. So looks like it's working as it is, unless I'm missing something?

AntonAM committed:
Fail on incorrectly signed PROXY line instead of ignoring it

AntonAM committed:
Adjust tests

AntonAM committed:
Add helper function for getting PROXY subject claim

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

strideynet commented:
> > The appropriate changes have been made for the ALPN Proxy - however - I believe you also need to add support for this into the SSH reverse tunnel proxy. See `lib/reversetunnel/transport.go` and follow the behaviour for when the address is the special constant `RemoteAuthServer` :)
> 
> It's automatically supported already, since in the `func (s *AuthProxyDialerService) HandleConnection()` after calling `s.dialAuthServe()` we still just get `net.conn` which was created over the tunnel under the hood. And we write PROXY header to it in the same way.
> 
> Here's example of debug logs with two clusters (both on ip `192.168.2.23), and we're trying to list resources with `tsh ls` from a different IP (192.168.2.15),
> 
> ```
> // Logs on the root cluster "tele2.port"
> Dec  8 12:48:42.946884 [AUTH]      DEBU DBGG: Listing resources on Auth server, peer address: 192.168.2.15:57393 auth/auth_with_roles.go:1198
> Dec  8 12:48:42.977950 [PROXY:SER] DEBU Requesting connection to @remote-auth-server [] in remote cluster. cluster:tele.port reversetunnel/remotesite.go:900
> 
> // Continuation of logs on the leaf cluster "tele.port"
> Dec  8 12:48:42.982045 DEBU [PROXY:AGE] Successfully dialed to @remote-auth-server "", start proxying. cluster:tele2.port reversetunnel/transport.go:306
> Dec  8 12:48:42.990815 DEBU [AUTH]      ClientCertPool -> cert(tele2.port issued by tele2.port:201604638341962741183117029050652578849) auth/middleware.go:676
> Dec  8 12:48:42.990956 DEBU [AUTH]      ClientCertPool -> cert(tele2.port issued by tele2.port:209682280188038834517368231345398959698) auth/middleware.go:676
> Dec  8 12:48:42.991061 DEBU [AUTH:1]    Server certificate cert(1b023d35-9484-48b3-b29d-788264c8de50.tele.port issued by tele.port:243357613536743763667664328504167968797). auth/middleware.go:312
> Dec  8 12:48:43.019890 DEBU             Mapped roles [access editor] of remote user "teleport2-admin" to local roles [access] and traits map[aws_role_arns:[] db_names:[] db_users:[] kubernetes_groups:[] kubernetes_users:[] logins:[root ubuntu ec2-user -teleport-internal-join] windows_logins:[]]. services/access_checker.go:447
> Dec  8 12:48:43.020701 DEBU             Skipping login -teleport-internal-join, not a valid Unix login. services/role.go:265
> Dec  8 12:48:43.021186 DEBU [AUTH]      DBGG: Listing resources on Auth server, peer address: 192.168.2.15:57394 auth/auth_with_roles.go:1198
> ```
> 
> If I disable sending out PROXY header in `func (s *AuthProxyDialerService) HandleConnection()` we don't get correct client IP on the remote auth server. So looks like it's working as it is, unless I'm missing something?

So to clarify, if the user connects via SSH rather than using TLS ALPN, their IP will be correctly propagated ?

AntonAM commented:
> So to clarify, if the user connects via SSH rather than using TLS ALPN, their IP will be correctly propagated ?

Yeah, thanks (and for the chat in slack), I see where is the problem now and it does not work through the SSH path 🤦 

What do you think about about handling it in a separate PR, based on this one (I'll change name of this one to reflect that it's only ALPN part)? I'd rather keep PRs smaller if possible, currently it's just stepping stones before we can add IP pinning checks enforcement, so it shouldn't matter if it's not all in one piece. It can also be related to the IP propagation to node servers, it also deals with reverse tunnels.

strideynet commented:
> > So to clarify, if the user connects via SSH rather than using TLS ALPN, their IP will be correctly propagated ?
> 
> Yeah, thanks (and for the chat in slack), I see where is the problem now and it does not work through the SSH path 🤦
> 
> What do you think about about handling it in a separate PR, based on this one (I'll change name of this one to reflect that it's only ALPN part)? I'd rather keep PRs smaller if possible, currently it's just stepping stones before we can add IP pinning checks enforcement, so it shouldn't matter if it's not all in one piece. It can also be related to the IP propagation to node servers, it also deals with reverse tunnels.

This is acceptable to me !

AntonAM commented:
> This is acceptable to me !

Great 👍  I've changed title of this PR. Everything else seems to be addressed.

jentfoo reviewed (commented):
Changes in general look good, just one question

AntonAM commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+			}
+
+			// If current proxy line was signed and verified, it takes precedence over new not signed proxy line
+			if proxyLine != nil && proxyLine.IsVerified {
```
If we're here it means new proxy line wasn't verified (because we're continuing to the next iteration if it's verified). So if there's already exists current proxy line, which was parsed in previous iteration, and it's signed, we do not overwrite it with new not-signed proxy line, but just continue. Let me know if I'm missing something or comment needs to be amended to be more clear.

jentfoo replied:
> If we're here it means new proxy line wasn't verified

Understood.  My original question came from a concern on if we need to be referencing only the first verified proxy line, or if it's ok to replace an existing verified proxy line with another verified proxy line (thus at the end of the loop referencing the last verified proxy line).  I understand now that it does not matter, and instead we are just trying to keep from replacing a verified line with a non-verified one.

Sorry for the confusion, thank you!

strideynet reviewed (approved):
Looks good - nice work !

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

espadolini commented on lib/jwt/jwt.go:
```diff
@@ -218,6 +218,33 @@ func (k *Key) SignAzureToken(claims AzureTokenClaims) (string, error) {
+const expirationPROXY = time.Second * 60
+
+func getPROXYSubjectClaim(sourceAddress, destinationAddress string) string {
+	return fmt.Sprintf("%s/%s", sourceAddress, destinationAddress)
```
Are we 100% sure there won't be any `/` in the addresses? And should we add some prefix to indicate that this is intended to be a PROXY source+dest claim?

espadolini replied:
Also, can't we have source addr and dest addr in separate (custom) claims, and maybe have a static string for the Subject instead?

strideynet replied:
> 

I did make a suggestion that the `sub` and `aud` fields could actually be ideal for this.

AntonAM replied:
I've changed it 👍 , addresses have their own fields now - source address goes to `sub` and destination address to `aud`

espadolini commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+}
+
+func GetSignedPROXYHeader(sourceAddress, destinationAddress net.Addr, clusterName string, signingCert []byte, signer PROXYSigner) ([]byte, error) {
+	sAddr := getTCPAddr(sourceAddress)
+	dAddr := getTCPAddr(destinationAddress)
+	if sAddr.Port < 0 || dAddr.Port < 0 {
```
If `getTCPAddr` can be fallible, make it return an error, IMO.

AntonAM replied:
I feel like code is a bit more compact and even clearer like this, would prefer to leave it like this, if it's ok with you.

espadolini commented on lib/multiplexer/multiplexer.go:
```diff
@@ -246,44 +256,145 @@ func (m *Mux) detectAndForward(conn net.Conn) {
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
```
Can we hoist this out of the loop? I'm not sure we have a TTL caching layer if caching is disabled, for instance, so this could end up hitting the backend multiple times.

We should only fetch the one CA we need on demand, similarly to how it's done to authenticate connections, IMO. It's not unheard for big setups to have hundreds (or thousands) of leaf clusters, and this would fetch and unmarshal all of them on each connection.

Besides, why would we trust a leaf cluster's word on an inbound connection's source?

AntonAM replied:
I see, didn't know there are setups with that many leaf clusters 🤯 I'll look into refactoring it into getting single required CA 🤔 
And yep, definitely good to hoist it out of cycle so we guaranteed to only call it one time.

>Besides, why would we trust a leaf cluster's word on an inbound connection's source?

It's a trusted cluster, so we trust it's CA anyway

espadolini replied:
With the way things are written here, a malicious _leaf_ cluster can produce a signature that will be trusted for the purpose of IP propagation.

A leaf cluster stores the parent's user CA (with a role map to translate role names) and not viceversa, but both parent and leaf store each other's host CAs.

AntonAM replied:
I've refactored host CA stuff - now we only try to get exact CA we know proxy cert comes from and we also check if we should trust  this cluster (basically if it's a root or local cluster).

espadolini commented on lib/service/service.go:
```diff
@@ -1674,9 +1677,10 @@ func (process *TeleportProcess) initAuthService() error {
+		EnableExternalProxyProtocol: cfg.Auth.EnableProxyProtocol,
+		Listener:                    listener,
+		ID:                          teleport.Component(process.id),
+		CertAuthoritiesGetter:       authCache,
```
Are we ok with disabling the entire feature if caching is disabled on the auth?

AntonAM replied:
Good point. So, I'll pass `authServer.Services` then, if cache is not initialized. 
What's the production use case for disabling cache btw?

espadolini replied:
I'd just pass `authServer` and let that figure it out.

AntonAM committed:
Pass authServer instead of authCache directly

We let authServer to figure out caching on its own - if caching is present
it will be used, if it's disabled, it will use services directly

AntonAM committed:
Use subject and audience jwt claims for source and dest addresses

AntonAM committed:
Check which clusters we can trust when verifying signed PROXY header

AntonAM committed:
Allow only one signed PROXY header

AntonAM committed:
User CertAuthorityGetter to verify if we can trust a cluster

jentfoo reviewed (approved):
I think the most recent changes look even better :+1:

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

espadolini reviewed (changes requested):
We can't expect agents to have cached access to the full set of certificate authorities; as we can make it so that the signed header is always added by the proxy that's local to the cluster (as it handles all connections from parent clusters, anyway), we should only ever need to trust signatures from the local cluster.

AntonAM committed:
Only allow signed PROXY headers from local cluster

AntonAM commented:
@espadolini just added check that we only allow signed headers from our local cluster.

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation

espadolini reviewed (approved):
Thank you for the quick fix for this!

AntonAM committed:
Merge branch 'master' into anton/proxy-ip-propagation
